src/AMDGPUOffload.cpp

```diff
@@ -0,0 +1,679 @@
+    }
+    return oss.str();
+}
+
+void do_reloc_64(char *addr, uintptr_t val) {
+    *((uint64_t*)addr) = (uint64_t)val;
+}
+
+void do_reloc_32(char *addr, uintptr_t val) {
+    internal_assert(val <= 0xffffffff);
+    *((uint32_t*)addr) = val;
+}
+
+void do_relocation(uint32_t fixup_offset, char *fixup_addr, uint32_t type,
+                   const Symbol *sym, uint32_t sym_offset, int32_t addend,
+                   Elf::Section &got) {
+    // Amdgpu relocations are specified in section 11.5 in
+    // the Amdgpu Application Binary Interface spec.
+
+    // Now we can define the variables from Table 11-5.
+    uint32_t S = sym_offset;
+    uint32_t P = fixup_offset;
+    intptr_t A = addend;
+
+    uint32_t G = got.contents_size();
+    for (const Relocation &r : got.relocations()) {
+        if (r.get_symbol() == sym) {
+            G = r.get_offset();
+            debug(2) << "Reusing G=" << G << " for symbol " << sym->get_name() << "\n";
+            break;
+        }
+    }
+
+    bool needs_got_entry = false;
+
+    switch (type) {
+    case R_AMDGPU_NONE:
+        break;
+    case R_AMDGPU_ABS32_LO:
+        do_reloc_32(fixup_addr, intptr_t((S + A) & 0xFFFFFFFF));
+        break;
+    case R_AMDGPU_ABS32_HI:
+        do_reloc_32(fixup_addr, intptr_t((S + A) >> 32));
+        break;
+    case R_AMDGPU_ABS64:
+        do_reloc_32(fixup_addr, intptr_t(S + A));
+        break;
+    case R_AMDGPU_REL32:
+        do_reloc_32(fixup_addr, intptr_t(S + A - P));
+        break;
+    case R_AMDGPU_REL64:
+        do_reloc_64(fixup_addr, intptr_t(S + A - P));
+        break;
+    case R_AMDGPU_ABS32:
+        do_reloc_32(fixup_addr, intptr_t(S + A));
+        break;
+    case R_AMDGPU_GOTPCREL:
+        do_reloc_32(fixup_addr, intptr_t(G + GOT + A - P));
+        break;
+    case R_AMDGPU_GOTPCREL32_LO:
+        do_reloc_32(fixup_addr, intptr_t((G + GOT + A - P) & 0xffffffff));
+        break;
+    case R_AMDGPU_GOTPCREL32_HI:
+        do_reloc_32(fixup_addr, intptr_t((G + GOT + A - P) >> 32));
+        break;
+    case R_AMDGPU_REL32_LO:
+        do_reloc_32(fixup_addr, intptr_t((S + A - P) & 0xffffffff));
+        break;
+    case R_AMDGPU_REL32_HI:
+        do_reloc_32(fixup_addr, intptr_t((S + A - P) >> 32));
+        break;
+    case R_AMDGPU_RELATIVE64:
+        break;
+    default:
+        internal_error << "Unhandled relocation type " << type << "\n";
+    }
+
+    if (needs_got_entry && G == got.contents_size()) {
+        debug(2) << "Adding GOT entry " << G << " for symbol " << sym->get_name() << "\n";
+        got.append_contents((uint64_t)0);
+        got.add_relocation(Relocation(R_AMDGPU_RELATIVE64, G, 0, sym));
+    }
+}
+
+class AMDGPULinker : public Linker {
+public:
+    uint32_t flags;
+
+    AMDGPULinker(const Target &target) {
+        if (target.has_feature(Target::AMDGPUGFX900)) {
+            flags = Elf::EF_AMDGPU_MACH_AMDGCN_GFX900;
+        } else {
+            flags = Elf::EF_AMDGPU_MACH_AMDGCN_GFX900;
```

Comment:
This needs to check for each target and set the flag artictecture field correctly. Suggest give an error if no valid target found. flags will also need to have  the xnack setting (EF_AMDGPU_XNACK) set which is likely another Target feature